crates/matrix-sdk-ui/src/timeline/inner/state.rs

```diff
@@ -491,6 +499,11 @@ impl TimelineInnerStateTransaction<'_> {
         }
     }
 
+    fn set_back_pagination_token(&mut self, token: String, event_id: Option<OwnedEventId>) {
+        trace!(token, ?event_id, "Adding back-pagination token");
+        self.meta.back_pagination_tokens.push((event_id, token));
```

Comment:
Is there any point in keeping back-pagination tokens without an event ID in positions other than the end of the list? Maybe pushing any token should remove the previously-last one if the first tuple element was `None`.

Reply:
I don't know, I'm not sure what is the purpose of keeping the event IDs :slightly_smiling_face:

Reply:
The idea is that we'll be able to remove chunks of the timeline from memory while retaining the ability of paginating them in again starting from the topmost remaining event.

Reply:
I see, then those might not be needed.

What do we do if we can't get the `event_id` of the latest event of a chunk? In practice that shouldn't happen but we still have that branch to handle. Since we keep track of all added events, we could probably use the ID of the previous event.

Reply:
Hm, I think handling this the same way as an empty chunk might be better as it's simpler. We'll have to be able to determine which timeline item a given token with event ID refers to, I think?

Reply:
That should be easy since we keep track of whether an event is visible or not.

Reply:
Alright, I'll just come to you if we get to that feature (discarding chunks) and I have questions about how to deal with invisible events 😄

Reply:
I found a corner case: what if the first chunk we receive is empty or has the latest event that fails to deserialize? We can't update the latest back-pagination token because there is none.

Reply:
Since back-pagination is already looping over multiple network requests, maybe we can rename `PaginationOptions::single_request` (e.g. `simple`) and make it loop until we either have an event item and pagination token to attach to it, or reached the start of the timeline?

Reply:
Oh wait, in your first comment are you saying we push tokens with `event_id == None`, but then we drop them when the next one arrives?

Reply:
We could also keep the `Option<String>` token in this PR (but maybe only use `None` in some cases?) and revisit later if you prefer.

Reply:
> Oh wait, in your first comment are you saying we push tokens with `event_id == None`, but then we drop them when the next one arrives?

Yes that's what I was saying, though you reached the logical conclusion I also had in my head that we really don't need the `Option` if we cover all the edge cases reasonably 😄